File: Sponsors/mw-streams-java/src/main/java/com/mwam/hackathon/operators/TumblingAverage.java
package com.mwam.hackathon.operators;

/**
 * Computes a tumbling average over some number of events.
 *
 * @param <I> Event type. Numeric, so
 */
public class TumblingAverage<I extends Number> extends Operator<I, Double> {
    private final long windowSize;
    private long count = 0;
    private double sum = 0.0;

    public TumblingAverage(long windowSize) {
        this.windowSize = windowSize;
    }

    @Override
    public void consume(I input) {
        sum += input.doubleValue();
        count += 1;

        if (count == windowSize) {
            double average = sum / count;
            count = 0;

            emit(average);
        }
    }

    @Override
    public void flush() {
        emit(this.sum / count);
        super.flush();
    }
}

File: Sponsors/mw_streams_python/mws/operators/tumbling_average.py
from numbers import Number

from mws.operators.operator import Operator


class TumblingAverage(Operator[Number, Number]):
    def __init__(self, window_size: int) -> None:
        super().__init__()
        self.window_size = window_size
        self.sum = 0.0
        self.count = 0

    def consume(self, event: Number) -> None:
        raise NotImplementedError("TODO implement me")
